titus_isolate/allocate/forecast_ip_cpu_allocator.py

```diff
@@ -68,41 +70,73 @@ def __init__(self,
         self.__config_manager = config_manager
         self.__cnt_rebalance_calls = 0
         self.__event_log_manager = None
-
+        self.__call_meta = None # track things __place_threads call
+
+
+    def __with_report(func):
+        @wraps(func)
+        def wrapped(inst, cpu: Cpu, workload_id: str, workloads: dict, cpu_usage: dict, instance_id: str):
+            inst.__call_meta = {}
+            report = lambda cpu_ : inst.report_cpu_event(
+                inst.__event_log_manager,
+                cpu_,
+                list(workloads.values()),
+                cpu_usage,
+                instance_id,
+                inst.__call_meta) 
+            try:
+                cpu = func(inst, cpu, workload_id, workloads, cpu_usage, instance_id)
+                report(cpu)
+                return cpu
+            except Exception as e:
+                report(cpu)
+                raise e
+        return wrapped
+
+    @__with_report
     def assign_threads(self, cpu: Cpu, workload_id: str, workloads: dict, cpu_usage: dict, instance_id: str) -> Cpu:
+        self.__call_meta['type'] = 'add'
```

Comment:
Should be ?
```self.__call_meta = {'type': 'add'}```

Reply:
If you do it this way you've got carryover metadata from previous invocations present.

Reply:
no, because thee function is wrapped in the `@__with_report` decorator